lib/logstash/codecs/cef.rb

```diff
@@ -58,28 +98,120 @@ def decode(data)
 
-    # TODO: Need to check that fields are set!
+    vendor = sanitize_header_field(event.sprintf(@vendor))
+    vendor = self.class.get_config["vendor"][:default] if vendor == ""
+
+    product = sanitize_header_field(event.sprintf(@product))
+    product = self.class.get_config["product"][:default] if product == ""
+
+    version = sanitize_header_field(event.sprintf(@version))
+    version = self.class.get_config["version"][:default] if version == ""
+
+    signature = sanitize_header_field(event.sprintf(@signature))
+    signature = self.class.get_config["signature"][:default] if signature == ""
+
+    name = sanitize_header_field(event.sprintf(@name))
+    name = self.class.get_config["name"][:default] if name == ""
+
+    # :sev is deprecated and therefore only considered if :severity equals the default setting or is invalid
+    severity = sanitize_severity(event, @severity)
+    if severity == self.class.get_config["severity"][:default]
+      # Use deprecated setting sev
+      severity = sanitize_severity(event, @sev)
+    end
 
-    # Signature, Name, and Sev should be set in the config, with ref to fields
     # Should also probably set the fields sent
-    header = ["CEF:0", "Elasticsearch", "Logstash", "1.0", @signature, @name, @sev].join("|")
-    values = @fields.map {|name| get_value(name, data)}.join(" ")
-    # values = values.map {|k,v| "#{k}=#{v}"}.join(" ")
-    @on_event.call(header + " " + values + "\n")
+    header = ["CEF:0", vendor, product, version, signature, name, severity].join("|")
+    values = @fields.map {|fieldname| get_value(fieldname, event)}.compact.join(" ")
+
+    @on_event.call(event, "#{header}|#{values}\n")
   end
 
   private
-  def get_value(name, event)
-    val = event[name]
+
+  # Escape pipes and backslashes in the header. Equal signs are ok.
+  # Newlines are forbidden.
+  def sanitize_header_field(value)
+    output = ""
+
+    value = value.to_s.gsub(/\r\n/, "\n")
+
+    value.each_char{|c|
+      case c
+      when "\\", "|"
+        output += "\\" + c
+      when "\n", "\r"
+        output += " "
+      else
+        output += c
+      end
+    }
+
+    return output
+  end
+
+  # Keys must be made up of a single word, with no spaces
+  # must be alphanumeric
+  def sanitize_extension_key(value)
+    value = value.to_s.gsub(/[^a-zA-Z0-9]/, "")
+    return value
+  end
+
+  # Escape equal signs in the extensions. Canonicalize newlines.
+  # CEF spec leaves it up to us to choose \r or \n for newline.
+  # We choose \n as the default.
+  def sanitize_extension_val(value)
+    output = ""
+
+    value = value.to_s.gsub(/\r\n/, "\n")
+
+    value.each_char{|c|
+      case c
+      when "\\", "="
+        output += "\\" + c
+      when "\n", "\r"
+        output += "\\n"
+      else
+        output += c
+      end
+    }
+
+    return output
+  end
+
+  def get_value(fieldname, event)
+    val = event[fieldname]
+
+    return nil if val.nil?
+
     case val
+    when Array
```

Comment:
since Array and Hash are the same, we can do 

```ruby
case val
when Array, Hash
  return "#{sanitize_extension_key(fieldname)}=#{sanitize_extension_val(val.to_json)}"
when ...
  ...
end
```